File: polyglot/ruby/dns_record_fetcher.rb
# frozen_string_literal: true

require 'resolv'
require 'socket'
require 'uri'
require 'timeout'

module Dmarcaudit
  module DnsRecordFetcher
    DEFAULT_TIMEOUT = 5 # seconds per query
    MAX_RETRIES = 3
    RETRY_DELAY = 0.1 # seconds between retries
    DEFAULT_DNS_SERVERS = ['8.8.8.8', '1.1.1.1', '208.67.222.222']

    class DnsError < StandardError; end

    class TimeoutError < DnsError; end

    class RecordNotFound < DnsError; end

    # Represents a parsed DNS record
    class ParsedRecord
      attr_reader :type, :name, :ttl, :content, :class

      def initialize(type:, name:, ttl: 300, content: '', class: 'IN')
        @type = type
        @name = name
        @ttl = ttl
        @content = content
        @class = class
      end

      def to_h
        { type: type, name: name, ttl: ttl, content: content, class: class }
      end

      def inspect
        "#<ParsedRecord #{type} #{name.inspect}> (#{content.inspect})"
      end
    end

    # Main DNS fetcher with retry logic and multiple server support
    class Client
      attr_reader :dns_servers, :default_timeout

      def initialize(dns_servers: DEFAULT_DNS_SERVERS, default_timeout: DEFAULT_TIMEOUT)
        @dns_servers = dns_servers
        @default_timeout = default_timeout
      end

      # Fetch a single record with retry and fallback logic
      def fetch(type:, name:, class: 'IN', timeout: nil)
        query_name = normalize_name(name)
        query_type = type.to_s.upcase
        query_class = class.to_s.upcase

        timeout ||= @default_timeout

        # Try each DNS server in order
        dns_servers.each do |server|
          result = attempt_query(server, query_name, query_type, query_class, timeout)
          return result if result
        end

        raise RecordNotFound, "Record not found after trying all servers: #{query_name} (#{query_type})"
      rescue TimeoutError => e
        raise TimeoutError, "Timeout fetching from all servers: #{e.message}"
      end

      # Fetch multiple records in parallel where possible
      def fetch_batch(types:, name:, class: 'IN', timeout: nil)
        results = {}

        types.each do |type|
          begin
            results[type] = fetch(type: type, name: name, class: class, timeout: timeout)
          rescue RecordNotFound => e
            results[type] = ParsedRecord.new(
              type: type,
              name: name,
              content: "#<#{e.class}: #{e.message}>"
            )
          end
        end

        results
      end

      # Fetch MX records (used for DMARC verification)
      def fetch_mx(name:, timeout: nil)
        fetch(type: 'MX', name: name, timeout: timeout).content.split("\n").map do |line|
          parts = line.strip.split
          raise RecordNotFound, "Malformed MX record: #{line}" if parts.size < 2

          { priority: parts[0].to_i, exchange: parts[1] }
        end.compact
      rescue RecordNotFound => e
        []
      end

      # Fetch TXT records (SPF, DMARC, DKIM)
      def fetch_txt(name:, timeout: nil)
        record = fetch(type: 'TXT', name: name, timeout: timeout)
        return [] if record.content.nil? || record.content.empty?

        record.content.split.map do |line|
          line.strip.presence || ''
        end.compact
      rescue RecordNotFound => e
        []
      end

      # Fetch AAAA records (IPv6)
      def fetch_aaaa(name:, timeout: nil)
        fetch(type: 'AAAA', name: name, timeout: timeout).content.split("\n").map do |ip|
          ip.strip.presence || ''
        end.compact
      rescue RecordNotFound => e
        []
      end

      # Fetch A records (IPv4)
      def fetch_a(name:, timeout: nil)
        fetch(type: 'A', name: name, timeout: timeout).content.split("\n").map do |ip|
          ip.strip.presence || ''
        end.compact
      rescue RecordNotFound => e
        []
      end

      # Fetch CNAME records
      def fetch_cname(name:, timeout: nil)
        fetch(type: 'CNAME', name: name, timeout: timeout).content.split("\n").map do |target|
          target.strip.presence || ''
        end.compact
      rescue RecordNotFound => e
        []
      end

      # Fetch SRV records (used for some DMARC configurations)
      def fetch_srv(name:, timeout: nil)
        fetch(type: 'SRV', name: name, timeout: timeout).content.split("\n").map do |line|
          parts = line.strip.split
          raise RecordNotFound, "Malformed SRV record: #{line}" if parts.size < 5

          {
            priority: parts[0].to_i,
            weight: parts[1].to_i,
            port: parts[2].to_i,
            target: parts[3]
          }
        end.compact
      rescue RecordNotFound => e
        []
      end

      # Fetch all records for a domain (useful for debugging)
      def fetch_all(name:, timeout: nil)
        {
          A: fetch_a(name: name, timeout: timeout),
          AAAA: fetch_aaaa(name: name, timeout: timeout),
          MX: fetch_mx(name: name, timeout: timeout),
          TXT: fetch_txt(name: name, timeout: timeout),
          CNAME: fetch_cname(name: name, timeout: timeout)
        }
      end

    private

      # Normalize the DNS query name (remove trailing dots, handle wildcards)
      def normalize_name(name)
        return name if name.nil? || name.empty?

        normalized = name.to_s.strip
        normalized = normalized.sub(/\.$/, '') unless normalized.end_with?('.')

        normalized
      end

      # Attempt a single DNS query with timeout and retry
      def attempt_query(server, name, type, class_, timeout)
        resolver = Resolv::DNS.new([server])
        begin
          result = resolver.getresources(name, type, class_)
          return parse_result(result, type, name) if result.any?

          # Check for NOERROR with empty answer (record exists but no data)
          response_code = resolver.last_response.code
          if response_code == 0 || response_code == Resolv::DNS::RCode::NOERROR
            return ParsedRecord.new(type: type, name: name, content: '')
          end

          # Check for NXDOMAIN
          if response_code == Resolv::DNS::RCode::NXDOMAIN
            raise RecordNotFound, "NXDOMAIN: #{name}"
          end

        rescue ResolvTimeoutError => e
          raise TimeoutError, "Query to #{server} timed out"
        rescue ResolvError => e
          raise DnsError, "Query error from #{server}: #{e.message}"
        end

        nil
      end

      # Parse the DNS response into a ParsedRecord
      def parse_result(resources, type, name)
        content_parts = resources.map do |rr|
          rr.to_s.split(' ').first(4).join(' ')
        end.join("\n")

        ParsedRecord.new(type: type, name: name, content: content_parts)
      rescue ResolvError => e
        raise DnsError, "Failed to parse DNS response: #{e.message}"
      end
    end

    # Convenience module methods that work with the default client
    module ClassMethods
      def fetch(type:, name:, class: 'IN', timeout: nil)
        Client.new.fetch(type: type, name: name, class: class, timeout: timeout)
      end

      def fetch_mx(name:, timeout: nil)
        Client.new.fetch_mx(name: name, timeout: timeout)
      end

      def fetch_txt(name:, timeout: nil)
        Client.new.fetch_txt(name: name, timeout: timeout)
      end

      def fetch_aaaa(name:, timeout: nil)
        Client.new.fetch_aaaa(name: name, timeout: timeout)
      end

      def fetch_a(name:, timeout: nil)
        Client.new.fetch_a(name: name, timeout: timeout)
      end

      def fetch_all(name:, timeout: nil)
        Client.new.fetch_all(name: name, timeout: timeout)
      end
    end

    # Extend the module with class methods for convenience
    include ClassMethods
  end
end

File: polyglot/rust/dns_record_fetcher.rs
use std::collections::{HashMap, HashSet};
use std::net::SocketAddr;
use std::time::Duration;
use thiserror::Error;
use tokio::net::UdpSocket;
use tokio::sync::mpsc;
use tokio::task::JoinSet;

/// Error types for DNS operations.
#[derive(Error, Debug)]
pub enum DnsError {
    #[error("DNS query failed: {0}")]
    QueryFailed(String),
    
    #[error("Timeout waiting for response from {0}")]
    Timeout(SocketAddr),
    
    #[error("Parse error: {0}")]
    ParseError(String),
    
    #[error("Network error: {0}")]
    Network(#[from] std::io::Error),
    
    #[error("Socket error: {0}")]
    Socket(#[from] tokio::net::UdpError),
}

/// Configuration for DNS resolver behavior.
#[derive(Debug, Clone)]
pub struct DnsConfig {
    /// List of DNS servers to query (for redundancy).
    pub dns_servers: Vec<SocketAddr>,
    
    /// Maximum retries per server before failing.
    pub max_retries: u32,
    
    /// Time to wait between retries.
    pub retry_delay: Duration,
    
    /// Total timeout for all queries combined.
    pub total_timeout: Duration,
}

impl Default for DnsConfig {
    fn default() -> Self {
        Self {
            dns_servers: vec![
                "8.8.8.8:53".parse().unwrap(),
                "1.1.1.1:53".parse().unwrap(),
                "208.67.222.222:53".parse().unwrap(),
            ],
            max_retries: 3,
            retry_delay: Duration::from_secs(1),
            total_timeout: Duration::from_secs(30),
        }
    }
}

/// A single DNS response from one server.
#[derive(Debug)]
pub struct DnsResponse {
    pub server: SocketAddr,
    pub records: Vec<DnsRecord>,
    pub query_time_ms: u64,
    pub truncated: bool,
}

impl Default for DnsResponse {
    fn default() -> Self {
        Self {
            server: "0.0.0.0".parse().unwrap(),
            records: Vec::new(),
            query_time_ms: 0,
            truncated: false,
        }
    }
}

/// A parsed DNS record with metadata.
#[derive(Debug)]
pub struct DnsRecord {
    pub name: String,
    pub rtype: RecordType,
    pub ttl: u32,
    pub data: String,
    pub source_server: SocketAddr,
}

impl Default for DnsRecord {
    fn default() -> Self {
        Self {
            name: "unknown".to_string(),
            rtype: RecordType::Unknown,
            ttl: 3600,
            data: String::new(),
            source_server: "0.0.0.0".parse().unwrap(),
        }
    }
}

/// DNS record types we care about for DMARC/SPF/DKIM.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RecordType {
    MX,
    TXT,
    A,
    AAAA,
    CNAME,
    NS,
    SRV,
    Unknown,
}

impl Default for RecordType {
    fn default() -> Self {
        Self::Unknown
    }
}

/// Result of a DNS query attempt.
#[derive(Debug)]
pub struct QueryResult {
    pub success: bool,
    pub response: Option<DnsResponse>,
    pub error: Option<DnsError>,
    pub attempts: u32,
}

impl Default for QueryResult {
    fn default() -> Self {
        Self {
            success: false,
            response: None,
            error: None,
            attempts: 0,
        }
    }
}

/// The main DNS fetcher implementation.
pub struct DnsFetcher {
    config: DnsConfig,
    socket_pool: HashMap<SocketAddr, UdpSocket>,
}

impl Default for DnsFetcher {
    fn default() -> Self {
        let mut fetcher = Self::new(DnsConfig::default());
        // Pre-create sockets for all configured servers
        for server in &fetcher.config.dns_servers {
            let socket = tokio::net::UdpSocket::bind("0.0.0.0:0").await.unwrap_or_else(|_| {
                UdpSocket::from_std(std::net::UdpStream::connect(*server).unwrap().into())
            });
            fetcher.socket_pool.insert(*server, socket);
        }
        fetcher
    }
}

impl DnsFetcher {
    /// Create a new DNS fetcher with custom configuration.
    pub fn new(config: DnsConfig) -> Self {
        let mut fetcher = Self {
            config,
            socket_pool: HashMap::new(),
        };
        
        // Initialize sockets for each server
        for server in &config.dns_servers {
            if let Ok(socket) = tokio::net::UdpSocket::bind("0.0.0.0:0") {
                fetcher.socket_pool.insert(*server, socket);
            } else {
                // Fallback to connecting directly
                if let Ok(stream) = std::net::UdpStream::connect(*server) {
                    fetcher.socket_pool.insert(*server, UdpSocket::from_std(stream));
                }
            }
        }
        
        fetcher
    }

    /// Fetch a single record type for a domain.
    pub async fn fetch_record(
        &self,
        domain: &str,
        rtype: RecordType,
    ) -> QueryResult {
        let mut result = QueryResult::default();
        let query_name = format!("_{rtype:?}.{domain}");
        
        // Try each server in order
        for (attempt, server) in self.config.dns_servers.iter().enumerate() {
            if attempt >= self.config.max_retries as usize {
                break;
            }
            
            let socket = match self.socket_pool.get(server) {
                Some(sock) => sock.clone(),
                None => {
                    // Create a new socket for this server
                    if let Ok(sock) = tokio::net::UdpSocket::bind("0.0.0.0:0") {
                        self.socket_pool.insert(*server, sock);
                        sock
                    } else {
                        continue;
                    }
                }
            };

            // Build DNS query packet (simplified for TXT/MX)
            let mut packet = [0u8; 512];
            
            // Header: Transaction ID + Flags
            packet[0..4].copy_from_slice(&[0, 0, 0, 1]); // Standard query
            
            // Question section (simplified - just length)
            let qname_len = query_name.len();
            packet[4] = (qname_len / 256) as u8;
            packet[5] = (qname_len % 256) as u8;
            
            // Question type and class
            match rtype {
                RecordType::TXT => {
                    packet[6] = 16; // TXT record type
                    packet[7] = 1;  // IN class (Internet)
                }
                RecordType::MX => {
                    packet[6] = 15; // MX record type
                    packet[7] = 1;
                }
                _ => {
                    packet[6] = 16; // Default to TXT
                    packet[7] = 1;
                }
            }

            let offset = 8;
            
            // Send the query
            let send_start = tokio::time::Instant::now();
            match socket.send_to(&packet[offset..], *server).await {
                Ok(sent) => {
                    if sent > 0 {
                        result.attempts += 1;
                        
                        // Set up timeout for this attempt
                        let total_elapsed = send_start.elapsed();
                        let remaining = self.config.total_timeout.saturating_sub(total_elapsed);
                        
                        tokio::select! {
                            recv = socket.recv_from(&mut packet[offset..]) => {
                                match recv {
                                    Ok((len, addr)) => {
                                        result.success = true;
                                        result.response = Some(DnsResponse {
                                            server: *addr,
                                            records: self.parse_response(
                                                &packet[offset..offset + len],
                                                query_name.clone(),
                                                rtype,
                                            ),
                                            query_time_ms: send_start.elapsed().as_millis() as u64,
                                            truncated: packet[0] & 0x20 != 0,
                                        });
                                    }
                                    Err(e) => {
                                        result.error = Some(DnsError::QueryFailed(format!("recv error: {}", e)));
                                    }
                                }
                            }
                            _ = tokio::time::sleep(remaining) => {
                                result.error = Some(DnsError::Timeout(*server));
                            }
                        }
                    } else {
                        result.error = Some(DnsError::QueryFailed("empty response".to_string()));
                    }
                }
                Err(e) => {
                    if attempt < self.config.max_retries as usize - 1 {
                        tokio::time::sleep(self.config.retry_delay).await;
                        continue;
                    } else {
                        result.error = Some(DnsError::QueryFailed(format!("send error: {}", e)));
                    }
                }
            }
        }

        result
    }

    /// Parse the raw DNS response into records.
    fn parse_response(&self, data: &[u8], query_name: String, rtype: RecordType) -> Vec<DnsRecord> {
        let mut records = Vec::new();
        
        if data.len() < 12 {
            return records;
        }

        // Skip header (12 bytes) and question section
        let offset = 12;
        
        // Parse answer section
        while offset + 4 <= data.len() && data[offset] != 0 {
            if offset >= data.len() - 5 {
                break;
            }

            // Read name length
            let mut name_offset = offset;
            let mut name_len = (data[name_offset] as usize) << 8 | (data[name_offset + 1] as usize);
            
            if name_offset + 2 > data.len() {
                break;
            }

            // Read name components
            let mut name_parts: Vec<String> = Vec::new();
            let mut current_offset = offset + 2;
            
            while current_offset < data.len() && !data[current_offset] == 0 {
                if current_offset >= data.len() - 1 {
                    break;
                }
                
                let part_len = (data[current_offset] as usize) << 8 | (data[current_offset + 1] as usize);
                if current_offset + 2 + part_len > data.len() {
                    break;
                }
                
                name_parts.push(String::from_utf8_lossy(&data[current_offset + 2..current_offset + 2 + part_len]).to_string());
                current_offset += 2 + part_len;
            }

            let full_name = if !name_parts.is_empty() {
                format!("{}.{}", name_parts.join("."), query_name)
            } else {
                query_name.clone()
            };

            offset = current_offset;

            // Read type and class (skip for now, we already know rtype)
            let _rtype_u16 = u16::from_le_bytes([data[offset], data[offset + 1]]);
            let _class_u16 = u16::from_le_bytes([data[offset + 2], data[offset + 3]]);
            offset += 4;

            // Read TTL (skip, use default)
            let _ttl = u32::from_be_bytes([data[offset], data[offset + 1], data[offset + 2], data[offset + 3]]);
            offset += 4;

            // Read record length
            if offset >= data.len() {
                break;
            }
            
            let rlen = (data[offset] as usize) << 8 | (data[offset + 1] as usize);
            offset += 2;

            // Extract the actual record data
            let record_data = if offset + rlen <= data.len() {
                String::from_utf8_lossy(&data[offset..offset + rlen]).to_string()
            } else {
                String::from_utf8_lossy(&data[offset..]).to_string()
            };

            // Only include records that match our expected type or are interesting
            if matches!(rtype, RecordType::TXT) || 
               record_data.starts_with("v=DMARC1") ||
               record_data.starts_with("v=spf1") {
                records.push(DnsRecord {
                    name: full_name,
                    rtype: rtype.clone(),
                    ttl: 3600, // Default TTL
                    data: record_data.trim().to_string(),
                    source_server: "0.0.0.0".parse().unwrap(),
                });
            }

            offset += rlen;
        }

        records
    }

    /// Fetch all relevant DMARC/SPF/DKIM records for a domain.
    pub async fn fetch_all_records(&self, domain: &str) -> HashMap<RecordType, QueryResult> {
        let mut results = HashMap::new();

        // MX record (always check first - critical for mail routing)
        results.insert(RecordType::MX, self.fetch_record(domain, RecordType::MX).await);

        // TXT records for DMARC and SPF
        results.insert(RecordType::TXT, self.fetch_record(domain, RecordType::TXT).await);

        // DKIM selectors (common ones)
        let dkim_selectors = ["default", "s1", "s2"];
        for selector in &dkim_selectors {
            let selector_domain = format!("{}._domainkey.{}", selector, domain);
            results.insert(RecordType::TXT, self.fetch_record(&selector_domain, RecordType::TXT).await);
        }

        // A record (basic connectivity check)
        results.insert(RecordType::A, self.fetch_record(domain, RecordType::A).await);

        results
    }

    /// Fetch records with a total timeout.
    pub async fn fetch_all_with_timeout(&self, domain: &str, timeout: Duration) -> HashMap<RecordType, QueryResult> {
        let mut results = HashMap::new();
        let start = tokio::time::Instant::now();

        // Spawn concurrent queries
        let mut handles = JoinSet::new();
        
        // MX + TXT (most important)
        handles.spawn(self.fetch_record(domain, RecordType::MX));
        handles.spawn(self.fetch_record(domain, RecordType::TXT));
        
        // DKIM selectors
        for selector in &["default", "s1"] {
            let sel_domain = format!("{}._domainkey.{}", selector, domain);
            handles.spawn(self.fetch_record(&sel_domain, RecordType::TXT));
        }

        while let Some(res) = handles.join_next().await {
            if start.elapsed() < timeout {
                match res {
                    Ok((rtype, result)) => results.insert(rtype, result),
                    Err(e) => eprintln!("Query error: {}", e),
                }
            } else {
                break;
            }
        }

        // Fill in any missing with defaults
        for rtype in &[RecordType::MX, RecordType::TXT] {
            if !results.contains_key(rtype) {
                results.insert(*rtype, QueryResult {
                    success: false,
                    response: None,
                    error: Some(DnsError::Timeout(timeout)),
                    attempts: 0,
                });
            }
        }

        results
    }

    /// Get the effective DNS servers being used.
    pub fn dns_servers(&self) -> &[SocketAddr] {
        &self.config.dns_servers
    }

    /// Set custom DNS servers.
    pub fn set_dns_servers(&mut self, servers: Vec<SocketAddr>) {
        self.config.dns_servers = servers;
        
        // Re-initialize sockets
        self.socket_pool.clear();
        for server in &self.config.dns_servers {
            if let Ok(socket) = tokio::net::UdpSocket::bind("0.0.0.0:0") {
                self.socket_pool.insert(*server, socket);
            }
        }
    }

    /// Get the current configuration.
    pub fn config